sigstore/_internal/oidc.py

```diff
@@ -54,3 +65,194 @@ def __init__(self, identity_token: str) -> None:
+            "code_challenge_method": "S256",
+            "state": self.state,
+            "nonce": self.nonce,
+        }
+
+    def auth_request(self) -> str:
+        params = self.auth_request_params()
+        return "https://oauth2.sigstore.dev/auth/auth?" + urllib.parse.urlencode(params)
+
+    def enable_oob(self) -> None:
+        self.is_oob = True
+
+
+def get_identity_token() -> str:
+    """
+    Retrieve an OpenID Connect token from the Sigstore provider.
+
+    This function and the components that it relies on are based off of:
+    https://github.com/psteniusubi/python-sample
+    """
+
+    code: str
+    with RedirectServer() as server:
+        # Launch web browser
+        if not webbrowser.open(server.base_uri):
+            print(f"Your browser will now be opened to:\n{server.auth_request()}\n")
+        else:
+            server.enable_oob()
+            print(
+                f"Go to the following link in a browser:\n\n\t{server.auth_request()}"
+            )
+
+        def handle_requests(server: RedirectServer, done: threading.Event) -> None:
+            # Continue processing HTTP requests until authorization response is received or the user
+            # has manually supplied the code in the out-of-band case.
+            while server.active and not done.is_set():
+                server.handle_request()
+
+        done = threading.Event()
+        thread = threading.Thread(
+            target=handle_requests,
+            args=(
+                server,
+                done,
+            ),
+        )
+        thread.start()
+
+        if not server.is_oob:
+            # In the normal case, we should join immediately and block until our server gets the
+            # token
+            thread.join()
+
+            # When the thread terminates, the response cannot be None
+            server.auth_response = cast(dict, server.auth_response)
+            auth_error = server.auth_response.get("error")
+            if auth_error is not None:
+                raise IdentityError(
+                    f"Error response from auth endpoint: {auth_error[0]}"
+                )
+            code = server.auth_response["code"][0]
+        else:
+            # In the out-of-band case, we wait until the user provides the code
+            #
+            # Once that is done, we need to manually signal to the redirect server that it should
+            # stop and THEN join
+            code = input("Enter verification code: ")
+            done.set()
+            thread.join()
+
+    # Provide code to token endpoint
+    data = {
+        "grant_type": "authorization_code",
+        "redirect_uri": server.redirect_uri,
+        "code": code,
+        "code_verifier": server.code_verifier.decode("utf-8"),
+    }
+    auth = (
+        "sigstore",
+        "",  # Client secret
+    )
+    resp: requests.Response = requests.post(
+        "https://oauth2.sigstore.dev/auth/token",
```

Comment:
could the oauth2 url not be configurable?

Reply:
I think the way it should work (and how `cosign` does it) is that you provide an "issuer" via the command line (by default it's `oauth2.sigstore.dev/auth`). Then we query the `.well-known/openid-configuration` and use the information in there to figure out where the auth and token endpoints are.

I intend to follow up with a change to add that functionality but I wanted to just get it working for `oauth2.sigstore.dev` as a first pass.

Reply:
sure, that's fine. understand